From 2b117b43bb34d5da17d096504c518ab5eaa362f0 Mon Sep 17 00:00:00 2001
Message-ID: <2b117b43bb34d5da17d096504c518ab5eaa362f0.1747662482.git.michal.simek@amd.com>
From: Michal Simek <michal.simek@amd.com>
Date: Fri, 25 Apr 2025 09:54:45 +0200
Subject: [PATCH] riscv: Wire CPU timers

Our HW is supporting them and FDT model is not wiring them.

Signed-off-by: Michal Simek <michal.simek@amd.com>
---
 target/riscv/cpu.c | 25 +++++++++++++------------
 1 file changed, 13 insertions(+), 12 deletions(-)

diff --git a/target/riscv/cpu.c b/target/riscv/cpu.c
index 2685728f8a28..3bd69a961830 100644
--- a/target/riscv/cpu.c
+++ b/target/riscv/cpu.c
@@ -553,6 +553,18 @@ static void rv128_base_cpu_init(Object *obj)
 #endif
 }
 #else
+#ifndef CONFIG_USER_ONLY
+static uint64_t mb_v_rdtime(void *opaque)
+{
+    CPURISCVState *env = &RISCV_CPU(opaque)->env;
+    uint32_t time, timeh;
+
+    csr_ops[CSR_MCYCLE].read(env, CSR_MCYCLE, &time);
+    csr_ops[CSR_MCYCLEH].read(env, CSR_MCYCLEH, &timeh);
+    return (uint64_t) timeh << 32 | time;
+}
+#endif
+
 static void rv32_base_cpu_init(Object *obj)
 {
     CPURISCVState *env = &RISCV_CPU(obj)->env;
@@ -562,6 +574,7 @@ static void rv32_base_cpu_init(Object *obj)
     env->priv_ver = PRIV_VERSION_LATEST;
 #ifndef CONFIG_USER_ONLY
     set_satp_mode_max_supported(RISCV_CPU(obj), VM_1_10_SV32);
+    riscv_cpu_set_rdtime_fn(env, mb_v_rdtime, obj);
 #endif
 }
 
@@ -634,18 +647,6 @@ static void rv32_imafcu_nommu_cpu_init(Object *obj)
     cpu->cfg.pmp = true;
 }
 
-#ifndef CONFIG_USER_ONLY
-static uint64_t mb_v_rdtime(void *opaque)
-{
-    CPURISCVState *env = &RISCV_CPU(opaque)->env;
-    uint32_t time, timeh;
-
-    csr_ops[CSR_MCYCLE].read(env, CSR_MCYCLE, &time);
-    csr_ops[CSR_MCYCLEH].read(env, CSR_MCYCLEH, &timeh);
-    return (uint64_t) timeh << 32 | time;
-}
-#endif
-
 static void rv32_microblaze_v_cpu_init(Object *obj)
 {
 #ifndef CONFIG_USER_ONLY
-- 
2.43.0

